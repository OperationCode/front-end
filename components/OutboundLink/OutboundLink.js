import React from 'react';
import { bool, node, string } from 'prop-types';
import { gtag } from 'common/utils/thirdParty/gtag';
import ExternalLinkIcon from 'static/images/icons/FontAwesome/external-link-square-alt-solid.svg';
import ScreenReaderOnly from 'components/ScreenReaderOnly/ScreenReaderOnly';
import styles from './OutboundLink.module.css';

OutboundLink.propTypes = {
  // will report this label plus the URL from where it was clicked
  analyticsEventLabel: string.isRequired,
  children: node.isRequired,
  className: string,
  'data-testid': string,
  hasIcon: bool,
  href: string.isRequired,
};

OutboundLink.defaultProps = {
  className: undefined,
  'data-testid': undefined,
  hasIcon: true,
};

function OutboundLink({
  analyticsEventLabel,
  children,
  'data-testid': testID,
  className,
  hasIcon,
  href,
}) {
  const isNotMailToLink = !href.startsWith('mailto:');

  const trackOutboundLinkClick = () => {
    gtag.outboundLink(analyticsEventLabel, href);
  };

  return (
    <a
      className={className}
      data-testid={testID}
      href={href}
      onClick={trackOutboundLinkClick}
      rel={isNotMailToLink ? 'noopener noreferrer' : undefined}
      target={isNotMailToLink ? '_blank' : undefined}
    >
      {children}

      <ScreenReaderOnly>Opens in new window</ScreenReaderOnly>

      {hasIcon && <ExternalLinkIcon className={styles.externalLinkIcon} />}
    </a>
  );
}

export default OutboundLink;
