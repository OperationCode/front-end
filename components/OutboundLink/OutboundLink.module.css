.externalLinkIcon {
  fill: currentColor;
  left: 4px;
  top: 2px;
  padding-right: 0.25rem;
  position: relative;
  width: 13px;
}
