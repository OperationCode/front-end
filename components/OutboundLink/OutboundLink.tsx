import classNames from 'classnames';
import { gtag } from '@/common/utils/thirdParty/gtag';
import ExternalLinkIcon from '@/public/static/images/icons/FontAwesome/external-link-square-alt-solid.svg';
import { ScreenReaderOnly } from '@/components/ScreenReaderOnly/ScreenReaderOnly';

export interface OutboundLinkPropsType {
  /**
   * will report this label plus the URL from where it was clicked
   */
  analyticsEventLabel: string;
  /**
   * Url path for the link.
   */
  href: string;
  /**
   * Content to be rendered as the link.
   */
  children: React.ReactNode;
  /**
   * Name of style class to use.
   */
  className?: string;
  /**
   * Sets an id to the base element for testing.
   */
  'data-testid'?: string;
  /**
   * Adds an an icon to identify link is an external link.
   */
  hasIcon?: boolean;
}

export function OutboundLink({
  analyticsEventLabel,
  children,
  'data-testid': testID,
  className,
  hasIcon = true,
  href,
}: OutboundLinkPropsType) {
  const isNotMailToLink = !href.startsWith('mailto:');

  const trackOutboundLinkClick = () => {
    gtag.outboundLink(analyticsEventLabel, href);
  };

  return (
    <a
      className={classNames('inline-flex items-start', className)}
      data-testid={testID}
      href={href}
      onClick={trackOutboundLinkClick}
      rel={isNotMailToLink ? 'noopener noreferrer' : undefined}
      target={isNotMailToLink ? '_blank' : undefined}
    >
      {children}

      <ScreenReaderOnly>Opens in new window</ScreenReaderOnly>

      {hasIcon && <ExternalLinkIcon className="!fill-current relative my-0 mx-1 bottom-0 w-3" />}
    </a>
  );
}
