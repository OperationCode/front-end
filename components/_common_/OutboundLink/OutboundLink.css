.screenReaderOnly {
  border: 0;
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.externalLinkIcon {
  fill: currentColor;
  left: 5px;
  top: 2px;
  position: relative;
  width: 13px;
}
