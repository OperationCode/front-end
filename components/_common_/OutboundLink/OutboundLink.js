import React from 'react';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { withRouter } from 'next/router';
import ExternalLinkIcon from 'static/images/icons/FontAwesome/external-link-square-alt-solid.svg';
import styles from './OutboundLink.css';

OutboundLink.propTypes = {
  analyticsEventLabel: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  hasIcon: PropTypes.bool,
  href: PropTypes.string.isRequired,
  router: PropTypes.object.isRequired,
};

OutboundLink.defaultProps = {
  className: '',
  hasIcon: true,
};

function OutboundLink({ analyticsEventLabel, children, className, hasIcon, href, router }) {
  const linkContent = (
    <React.Fragment>
      <span className={styles.screenReaderOnly}>Opens in new window</span>
      {children}
      {hasIcon && <ExternalLinkIcon className={styles.externalLinkIcon} />}
    </React.Fragment>
  );

  if (process.env.NODE_ENV === 'production') {
    return (
      <ReactGA.OutboundLink
        className={className}
        eventLabel={`OUTBOUND [${analyticsEventLabel}] from ${router.route}`}
        rel="noopener noreferrer"
        target="_blank"
        to={href}
      >
        {linkContent}
      </ReactGA.OutboundLink>
    );
  }

  return (
    <a className={className} href={href} rel="noopener noreferrer" target="_blank">
      {linkContent}
    </a>
  );
}

export default withRouter(OutboundLink);
