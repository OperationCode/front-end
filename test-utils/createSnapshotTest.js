/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
