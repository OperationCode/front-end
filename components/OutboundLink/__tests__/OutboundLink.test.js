import { render } from '@testing-library/react';
import React from 'react';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import OutboundLink from '../OutboundLink';

describe('OutboundLink', () => {
  it('should render with required props', () => {
    createSnapshotTest(
      <OutboundLink analyticsEventLabel="Test" href="https://tests.com">
        Test
      </OutboundLink>,
    );
  });

  it('should render SVG when `hasIcon` is true', () => {
    const { container } = render(
      <OutboundLink
        analyticsEventLabel="Test"
        className="test-class"
        href="https://tests.com"
        hasIcon
      >
        Test
      </OutboundLink>,
    );

    expect(container.querySelector('svg')).not.toBeNull();
  });

  it('should not render SVG when `hasIcon` is false', () => {
    const { container } = render(
      <OutboundLink
        analyticsEventLabel="Test"
        className="test-class"
        href="https://tests.com"
        hasIcon={false}
      >
        Test
      </OutboundLink>,
    );

    expect(container.querySelector('svg')).toBeNull();
  });
});
