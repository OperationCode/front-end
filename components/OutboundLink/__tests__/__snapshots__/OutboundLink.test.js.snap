// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutboundLink should render with required props 1`] = `
<a
  href="https://tests.com"
  onClick={[Function]}
  rel="noopener noreferrer"
  target="_blank"
>
  Test
  <span
    className="ScreenReaderOnly"
    data-testid="SCREEN_READER_ONLY"
  >
    Opens in new window
  </span>
  <svg
    className="externalLinkIcon"
  />
</a>
`;
